Adding Redis as a database option to Cactus

masri2019 opened the pull request:
This PR contains the changes made to add Redis  as a database option to Cactus beside the currently used database, Kyoto Tycoon. The main changes are:

- Adding [Redis](https://github.com/redis/redis) and its C client library, [hiredis](https://github.com/redis/hiredis), as submodules to Cactus. I didn't use their Ubuntu packages since they are several versions behind the latest ones;
  - redis: [4.0.9](https://packages.ubuntu.com/bionic/redis-server) (the latest is [6.0.8](https://github.com/redis/redis/tags))
  - hiredis: [0.13](https://packages.ubuntu.com/bionic/libhiredis-dev) (the latest is [1.0.0](https://github.com/redis/hiredis/releases))

  (Maybe we can change it to use the ubuntu packages if new submodules make it difficult for users to build cactus but in that case we may not be able to use the latest versions)
- Adding python scripts for running a Redis server (or KT server) as a Toil Job.Service
- Adding a python test script ([pipeline/dbServerTest.py](https://github.com/ComparativeGenomicsToolkit/cactus/blob/redis/src/cactus/pipeline/dbServerTest.py)) to test launching the databases, saving and loading snapshot and shutting down
- Adding a test function to [tests/evolverTest.py](https://github.com/ComparativeGenomicsToolkit/cactus/blob/46baf25f200ea35bcb81e91bb0503f44175785d9/test/evolverTest.py#L316) for testing cactus with Redis
- Changed the HEAD of sonLib submodule to the [redis](https://github.com/ComparativeGenomicsToolkit/sonLib/tree/redis) branch of sonLib (We can change it whenever this branch is merged with sonLib master)

diekhans reviewed (changes requested):
we should not build redis a submodule.  It is a maintained software with packages available.  The reason for building koyto and having our own fork is that it was stagnant software that we had to modify to keep it compiled.

masri2019 commented:
@diekhans I see your point. In these recent commits I removed all stuff related to redis and hiredis submodules. I added the packages that has to be installed in the Dockerfile. @glennhickey About the Gitlab test I added `redis-server` and `libhiredis` to the `.gitlab-ci.yml` (and also to `.travis.yml` for passing Travis CI) and as you can see `test/evolverTest.py::TestCase::testEvolverRedisLocal` is passed successfully. Maybe we have to remove them from `.gitlab-ci.yml` and install in your image `quay.io/glennhickey/cactus-ci-base:latest` in the future.

glennhickey commented:
Outstanding.  This looks near-flawless to me.  Some mostly small issues
* `cactus-blast` shouldn't have a `--database` option as it never makes a database.  If the shared code it uses wants one, it can be just be hardcoded to ktserver internally.  
* Please PR your sonLib brach
* Not a big deal but would encourage you to clean up the commit history a bit with squash an `git rebase -i`.  
* I tried running without redis installed locally, and it just hangs forever when trying to run the command.  It would be better to detect this and send an error right away.
* I couldn't run the evolver tests on my desktop even after installing.  I get an error from what looks like redis means as a warning (same command does work with `--binariesMode docker` though):
```
	  File "/home/hickey/dev/cactus/venv41/lib/python3.6/site-packages/cactus/pipeline/redisServerControl.py", line 72, in runServer
	    raise RuntimeError("Unable to launch redis-server in time. Log: %s" % log)
	RuntimeError: Unable to launch redis-server in time. Log: 18880:C 26 Oct 10:47:11.761 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
	18880:C 26 Oct 10:47:11.761 # Redis version=4.0.9, bits=64, commit=00000000, modified=0, pid=18880, just started
	18880:C 26 Oct 10:47:11.761 # Configuration loaded
	18880:M 26 Oct 10:47:11.762 # You requested maxclients of 10000 requiring at least 10032 max file descriptors.
	18880:M 26 Oct 10:47:11.762 # Server can't set maximum open files to 10032 because of OS error: Operation not permitted.
	18880:M 26 Oct 10:47:11.762 # Current maximum open files is 4096. maxclients has been reduced to 4064 to compensate for low ulimit. If you need higher maxclients increase 'ulimit -n'.
	18880:M 26 Oct 10:47:11.762 * Running mode=standalone, port=19948.
	18880:M 26 Oct 10:47:11.762 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
	18880:M 26 Oct 10:47:11.762 # Server initialized
	18880:M 26 Oct 10:47:11.762 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
	18880:M 26 Oct 10:47:11.762 * Ready to accept connections
```

glennhickey commented:
Just noticed that my run with Docker hung shortly after the setup phase.  
```
boreale 2020-10-26 12:12:58,945 MainThread INFO toil-rt: 2020-10-26 12:12:58.945824: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 5a94a69e-c1be-44b9-943d-ea756f853dbc --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 cactus_workflow_flowerStats INFO '<st_kv_database_conf type="redis">
			<redis database_dir="fakepath" host="192.168.1.20" port="20648" />
		</st_kv_database_conf>
	' 0" in 0.8477 seconds and 62.4 M memory
boreale 2020-10-26 12:12:58,946 MainThread INFO toil-rt: 2020-10-26 12:12:58.946960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 8c0e61bc-3a4b-482a-ac50-0f13eb5e6686 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 set TERMINATE 1"
boreale 2020-10-26 12:12:59,458 MainThread INFO toil-rt: 2020-10-26 12:12:59.458341: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpus11oml2/e4a418c0-0b4e-4aff-8e50-6c7c388f02c5:/data --entrypoint /opt/cactus/wrapper.sh --name 87200bb0-6751-46e3-a1b0-408a71518f97 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 set TERMINATE 1" in 0.887 seconds and 63.3 M memory
boreale 2020-10-26 12:12:59,894 MainThread INFO toil-rt: 2020-10-26 12:12:59.894378: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 8c0e61bc-3a4b-482a-ac50-0f13eb5e6686 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 set TERMINATE 1" in 0.9435 seconds and 62.6 M memory
boreale 2020-10-26 12:13:43,726 MainThread INFO toil-rt: 2020-10-26 12:13:43.725960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqivlvuod/f92aa2be-fef6-4232-bc8a-20327b184f6e:/data --entrypoint /opt/cactus/wrapper.sh --name 43347803-c468-4571-b634-e42368d86500 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 get TERMINATE"
boreale 2020-10-26 12:13:43,729 MainThread INFO toil-rt: 2020-10-26 12:13:43.728960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqnyy7wbw/5549753c-7195-4f28-9971-4dc43dce216d:/data --entrypoint /opt/cactus/wrapper.sh --name dda0ba32-e558-4309-bc60-a0197cc506bd --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 get TERMINATE"
boreale 2020-10-26 12:13:44,660 MainThread INFO toil-rt: 2020-10-26 12:13:44.660309: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqnyy7wbw/5549753c-7195-4f28-9971-4dc43dce216d:/data --entrypoint /opt/cactus/wrapper.sh --name dda0ba32-e558-4309-bc60-a0197cc506bd --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 get TERMINATE" in 0.9218 seconds and 63.2 M memory
boreale 2020-10-26 12:13:44,721 MainThread INFO toil-rt: 2020-10-26 12:13:44.721519: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqivlvuod/f92aa2be-fef6-4232-bc8a-20327b184f6e:/data --entrypoint /opt/cactus/wrapper.sh --name 43347803-c468-4571-b634-e42368d86500 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 get TERMINATE" in 0.9844 seconds and 62.6 M memory
```
and then nothing for going on about an hour..

masri2019 commented:
> Outstanding. This looks near-flawless to me. Some mostly small issues
> 
> * `cactus-blast` shouldn't have a `--database` option as it never makes a database.  If the shared code it uses wants one, it can be just be hardcoded to ktserver internally.
> * Please PR your sonLib brach
> * Not a big deal but would encourage you to clean up the commit history a bit with squash an `git rebase -i`.
> * I tried running without redis installed locally, and it just hangs forever when trying to run the command.  It would be better to detect this and send an error right away.
> * I couldn't run the evolver tests on my desktop even after installing.  I get an error from what looks like redis means as a warning (same command does work with `--binariesMode docker` though):
> 
> ```
> 	  File "/home/hickey/dev/cactus/venv41/lib/python3.6/site-packages/cactus/pipeline/redisServerControl.py", line 72, in runServer
> 	    raise RuntimeError("Unable to launch redis-server in time. Log: %s" % log)
> 	RuntimeError: Unable to launch redis-server in time. Log: 18880:C 26 Oct 10:47:11.761 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
> 	18880:C 26 Oct 10:47:11.761 # Redis version=4.0.9, bits=64, commit=00000000, modified=0, pid=18880, just started
> 	18880:C 26 Oct 10:47:11.761 # Configuration loaded
> 	18880:M 26 Oct 10:47:11.762 # You requested maxclients of 10000 requiring at least 10032 max file descriptors.
> 	18880:M 26 Oct 10:47:11.762 # Server can't set maximum open files to 10032 because of OS error: Operation not permitted.
> 	18880:M 26 Oct 10:47:11.762 # Current maximum open files is 4096. maxclients has been reduced to 4064 to compensate for low ulimit. If you need higher maxclients increase 'ulimit -n'.
> 	18880:M 26 Oct 10:47:11.762 * Running mode=standalone, port=19948.
> 	18880:M 26 Oct 10:47:11.762 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
> 	18880:M 26 Oct 10:47:11.762 # Server initialized
> 	18880:M 26 Oct 10:47:11.762 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
> 	18880:M 26 Oct 10:47:11.762 * Ready to accept connections
> ```

@glennhickey Thanks for your feedback. I think it's better to focus on the errors then I'll do the others. The warnings that you see in the redis log also appears for me but it doesn't produce any errors! When it logs `Ready to accept connections` there shouldn't be any problem on the redis side I think. That's what I used for making sure the redis is running successfully [redisServerControl.py](https://github.com/ComparativeGenomicsToolkit/cactus/blob/7a37089533e98123a4ec0ab2a8e8d255dabcc6c1/src/cactus/pipeline/redisServerControl.py#L105)

masri2019 commented:
I guess I found the difference. I don't get this error
`Server can't set maximum open files to 10032 because of OS error: Operation not permitted.`
Instead I get this message
`Increased maximum number of open files to 1032 (it was originally set to 1024)`

masri2019 commented:
> Just noticed that my run with Docker hung shortly after the setup phase.
> 
> ```
> boreale 2020-10-26 12:12:58,945 MainThread INFO toil-rt: 2020-10-26 12:12:58.945824: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 5a94a69e-c1be-44b9-943d-ea756f853dbc --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 cactus_workflow_flowerStats INFO '<st_kv_database_conf type="redis">
> 			<redis database_dir="fakepath" host="192.168.1.20" port="20648" />
> 		</st_kv_database_conf>
> 	' 0" in 0.8477 seconds and 62.4 M memory
> boreale 2020-10-26 12:12:58,946 MainThread INFO toil-rt: 2020-10-26 12:12:58.946960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 8c0e61bc-3a4b-482a-ac50-0f13eb5e6686 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 set TERMINATE 1"
> boreale 2020-10-26 12:12:59,458 MainThread INFO toil-rt: 2020-10-26 12:12:59.458341: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpus11oml2/e4a418c0-0b4e-4aff-8e50-6c7c388f02c5:/data --entrypoint /opt/cactus/wrapper.sh --name 87200bb0-6751-46e3-a1b0-408a71518f97 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 set TERMINATE 1" in 0.887 seconds and 63.3 M memory
> boreale 2020-10-26 12:12:59,894 MainThread INFO toil-rt: 2020-10-26 12:12:59.894378: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpd5d35ye_/66e5a48f-2f97-4f1e-bb2f-6f3e8969271f:/data --entrypoint /opt/cactus/wrapper.sh --name 8c0e61bc-3a4b-482a-ac50-0f13eb5e6686 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 set TERMINATE 1" in 0.9435 seconds and 62.6 M memory
> boreale 2020-10-26 12:13:43,726 MainThread INFO toil-rt: 2020-10-26 12:13:43.725960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqivlvuod/f92aa2be-fef6-4232-bc8a-20327b184f6e:/data --entrypoint /opt/cactus/wrapper.sh --name 43347803-c468-4571-b634-e42368d86500 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 get TERMINATE"
> boreale 2020-10-26 12:13:43,729 MainThread INFO toil-rt: 2020-10-26 12:13:43.728960: Running the command: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqnyy7wbw/5549753c-7195-4f28-9971-4dc43dce216d:/data --entrypoint /opt/cactus/wrapper.sh --name dda0ba32-e558-4309-bc60-a0197cc506bd --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 get TERMINATE"
> boreale 2020-10-26 12:13:44,660 MainThread INFO toil-rt: 2020-10-26 12:13:44.660309: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqnyy7wbw/5549753c-7195-4f28-9971-4dc43dce216d:/data --entrypoint /opt/cactus/wrapper.sh --name dda0ba32-e558-4309-bc60-a0197cc506bd --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 35887 -h 192.168.1.20 get TERMINATE" in 0.9218 seconds and 63.2 M memory
> boreale 2020-10-26 12:13:44,721 MainThread INFO toil-rt: 2020-10-26 12:13:44.721519: Successfully ran: "docker run --interactive --net=host --log-driver=none -u 1000:1000 -v /home/hickey/dev/cactus/work/node-8884765a-90ac-46be-b91e-45d29445f789-e1427db94b60421098ed79e7d63f0687/tmpqivlvuod/f92aa2be-fef6-4232-bc8a-20327b184f6e:/data --entrypoint /opt/cactus/wrapper.sh --name 43347803-c468-4571-b634-e42368d86500 --rm quay.io/comparative-genomics-toolkit/cactus:7a37089533e98123a4ec0ab2a8e8d255dabcc6c1 redis-cli -p 20648 -h 192.168.1.20 get TERMINATE" in 0.9844 seconds and 62.6 M memory
> ```
> 
> and then nothing for going on about an hour..

@glennhickey Is it possible to share the docker image you are using for this?

masri2019 commented:
It seems that the problem is redis opens as many files as the maximum number of clients that can connect to the redis server simultaneously. In the recent commit I restricted `--maxclients 200` 
200 should be large enough for cactus threads and small enough for the OS restriction on max number of open files. I think it should fix the error you saw in the local mode (and maybe for the docker mode?). @glennhickey

glennhickey commented:
i found two issues on my end.  First, I didn't run `make clean` after checking out your branch, so my sonlib wasn't fully redis-supporting.  Running on the latest commit gave me an error right away that pointed me to that (but I can't rule out that I just didn't see it last time).  Second, I recently added a little hack to cactus to print memory usage of each command when `CACTUS_LOG_MEMORY` is defined.  It doesn't work with servers though, and I just pushed a commit to turn it off for `redis-server`.  Because I had that variant set, I was seeing different behaviour...

masri2019 commented:
@glennhickey I rebased this PR to clean the commit history and also did some other changes:

- I removed `--database` argument from `cactus-blast` and it's hard-coded to `kyoto_tycoon` [here ](https://github.com/ComparativeGenomicsToolkit/cactus/blob/daf9961193d9a43311c432edd1c90b57968f585a/src/cactus/blast/cactus_blast.py#L69)

- Since you merged the sonLib PR I [updated sonLib HEAD to the latest commit on the master branch](https://github.com/ComparativeGenomicsToolkit/cactus/pull/338/commits/d8c36bc49878355697220f698a374213c801f5ae)

- Before launching a redis server I [checked the version of `redis-server` through `cactus_call` ](https://github.com/ComparativeGenomicsToolkit/cactus/pull/338/commits/daf9961193d9a43311c432edd1c90b57968f585a)to make sure if `redis-server` is installed. This way if it is not installed cactus will raise error instantly.

Please let know if there is anything else to be done.

glennhickey commented:
Awesome, thanks!!  Look forward to trying it out.  I have some pangenome data I'm running again and again, and will report the effects on runtime for that if I find any.

masri2019 commented:
> Hi Mobin, would it be difficult to leave redis as an option? So if it's not installed on the system, sonlib doesn't build in support and cactus still works so long as --database redis isn't used? I'd assumed this was how it was currently working via that HAVE_REDIS flag or whatever it is.
> […](#)
> On Thu, Nov 5, 2020 at 12:20 PM Mobin Asri ***@***.***> wrote: @thiagogenez <https://github.com/thiagogenez> Thanks for your comment. It is already mentioned in the README you can use the Dockerfile <https://github.com/ComparativeGenomicsToolkit/cactus/blob/89058e4f046048bed25a56537b5803349df9cf52/Dockerfile#L4> as a guide to see how all dependencies are installed with apt on Ubuntu. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[d3d6361#commitcomment-43911229](https://github.com/ComparativeGenomicsToolkit/cactus/commit/d3d6361914091a97a56946f56763ead42e84ff90#commitcomment-43911229)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAG373US2T64TF6KY3J6T4LSOLNEBANCNFSM4TKMYVNA> .

@glennhickey Yes, It would be easy to do that. Just a few lines should be added to [include.mk](https://github.com/ComparativeGenomicsToolkit/cactus/blob/89058e4f046048bed25a56537b5803349df9cf52/include.mk#L50) to check the existence of `hiredis` I'll do a small PR for that today.
